src/System.Runtime/tests/System/LazyTests.cs

```diff
@@ -103,17 +106,283 @@ public static void ToString_DoesntForceAllocation()
+            CheckException(typeof(MyException), new Lazy<int>(() => { throw new MyException(99); }, LazyThreadSafetyMode.None));
+            CheckException(typeof(MyException), new Lazy<int>(() => { throw new MyException(99); }, LazyThreadSafetyMode.PublicationOnly));
+
+            CheckException(typeof(TargetInvocationException), new Lazy<ExceptionInCtor>());
+            CheckException(typeof(TargetInvocationException), new Lazy<ExceptionInCtor>(true));
+            CheckException(typeof(TargetInvocationException), new Lazy<ExceptionInCtor>(false));
+            CheckException(typeof(TargetInvocationException), new Lazy<ExceptionInCtor>(LazyThreadSafetyMode.ExecutionAndPublication));
+            CheckException(typeof(TargetInvocationException), new Lazy<ExceptionInCtor>(LazyThreadSafetyMode.None));
+            CheckException(typeof(TargetInvocationException), new Lazy<ExceptionInCtor>(LazyThreadSafetyMode.PublicationOnly));
+
+            CheckException(typeof(MyException), new Lazy<ExceptionInCtor>(() => new ExceptionInCtor(99)));
+            CheckException(typeof(MyException), new Lazy<ExceptionInCtor>(() => new ExceptionInCtor(99), true));
+            CheckException(typeof(MyException), new Lazy<ExceptionInCtor>(() => new ExceptionInCtor(99), false));
+            CheckException(typeof(MyException), new Lazy<ExceptionInCtor>(() => new ExceptionInCtor(99), LazyThreadSafetyMode.ExecutionAndPublication));
+            CheckException(typeof(MyException), new Lazy<ExceptionInCtor>(() => new ExceptionInCtor(99), LazyThreadSafetyMode.None));
+            CheckException(typeof(MyException), new Lazy<ExceptionInCtor>(() => new ExceptionInCtor(99), LazyThreadSafetyMode.PublicationOnly));
+        }
+
+        private static void SameException<T>(Lazy<T> x)
+        {
+            Exception first = null;
+            try
+            {
+                var _ = x.Value;
+            }
+            catch (Exception thrown1)
+            {
+                first = thrown1;
+            }
+            Assert.NotNull(first);
+
+            try
+            {
+                var _ = x.Value;
+            }
+            catch (MyException thrown2)
+            {
+                Assert.Same(first, thrown2);
+            }
+        }
+
+        private static void DifferentException<T>(Lazy<T> x)
+        {
+            Exception first = null;
+            try
+            {
+                var _ = x.Value;
+            }
+            catch (Exception thrown1)
+            {
+                first = thrown1;
+            }
+            Assert.NotNull(first);
+
+            Exception second = null;
+            try
+            {
+                var _ = x.Value;
+            }
+            catch (Exception thrown2)
+            {
+                second = thrown2;
+            }
+            Assert.NotNull(second);
+
+            Assert.NotEqual(first, second);
+        }
```

Comment:
This can similarly be condensed, and just use Assert.NotSame instead of Assert.Same.